File: Python/ScrambleString.py
from functools import lru_cache


class Solution:
    def isScramble(self, s1: str, s2: str) -> bool:
        # Recursion on substrings with memo. For every split point i, s1 is a scramble of s2
        # if either (no swap) the left i chars and right n-i chars both match, or (swap) s1's
        # left i matches s2's right i and s1's right n-i matches s2's left n-i.
        # Anagram pre-check prunes hopeless branches. O(n^4) states*work; n <= 30.
        @lru_cache(maxsize=None)
        def solve(a: str, b: str) -> bool:
            if a == b:
                return True
            if sorted(a) != sorted(b):
                return False
            n = len(a)
            for i in range(1, n):
                if solve(a[:i], b[:i]) and solve(a[i:], b[i:]):
                    return True
                if solve(a[:i], b[n - i:]) and solve(a[i:], b[:n - i]):
                    return True
            return False

        return solve(s1, s2)


if __name__ == '__main__':
    s = Solution()
    print(s.isScramble("great", "rgeat"))  # True
    print(s.isScramble("abcde", "caebd"))  # False
    print(s.isScramble("a", "a"))          # True
    print(s.isScramble("abc", "bca"))      # True

File: Go/ScrambleString.go
//go:build ignore

package main

import "fmt"

func isScramble(s1, s2 string) bool {
	memo := make(map[string]bool)

	var solve func(a, b string) bool
	solve = func(a, b string) bool {
		if a == b {
			return true
		}
		key := a + "|" + b
		if v, ok := memo[key]; ok {
			return v
		}
		// Anagram check via character counts.
		var cnt [26]int
		for i := 0; i < len(a); i++ {
			cnt[a[i]-'a']++
			cnt[b[i]-'a']--
		}
		for _, c := range cnt {
			if c != 0 {
				memo[key] = false
				return false
			}
		}
		n := len(a)
		for i := 1; i < n; i++ {
			if solve(a[:i], b[:i]) && solve(a[i:], b[i:]) {
				memo[key] = true
				return true
			}
			if solve(a[:i], b[n-i:]) && solve(a[i:], b[:n-i]) {
				memo[key] = true
				return true
			}
		}
		memo[key] = false
		return false
	}

	return solve(s1, s2)
}

func main() {
	fmt.Println(isScramble("great", "rgeat")) // true
	fmt.Println(isScramble("abcde", "caebd")) // false
	fmt.Println(isScramble("a", "a"))         // true
	fmt.Println(isScramble("abc", "bca"))     // true
}
